File: alert_app/android/app/src/main/java/io/flutter/plugins/alert_app/SmsReceiver.java
package io.flutter.plugins.alert_app;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SmsMessage;
import android.util.Log;
import java.util.HashMap;
import java.util.Map;
import com.example.alert_new.MainActivity;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import android.app.PendingIntent;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;

public class SmsReceiver extends BroadcastReceiver {
    private static final String TAG = "SmsReceiver";
    
    private static final String[] UPI_KEYWORDS = {
        "credited", "received", "paid", "upi", "transaction", "₹", "rs.", "amount"
    };
    
    private static final String[] UPI_SENDERS = {
        "GOOGLEPAY", "PHONEPE", "PAYTM", "BHIMUPI", "AMAZONPAY", "MOBIKWIK", "FREECHARGE"
    };

    @Override
    public void onReceive(Context context, Intent intent) {
        if ("android.provider.Telephony.SMS_RECEIVED".equals(intent.getAction())) {
            Bundle bundle = intent.getExtras();
            if (bundle != null) {
                Object[] pdus = (Object[]) bundle.get("pdus");
                if (pdus != null) {
                    for (Object pdu : pdus) {
                        SmsMessage sms = SmsMessage.createFromPdu((byte[]) pdu);
                        String sender = sms.getDisplayOriginatingAddress();
                        String message = sms.getDisplayMessageBody();
                        
                        Log.d(TAG, "SMS from: " + sender + ", Message: " + message);
                        
                        if (isPaymentSms(sender, message)) {
                            Log.d(TAG, "Payment SMS detected");
                            sendSmsToFlutter(sender, message);
                            showNotification(context, "Payment Alert", "New UPI payment detected: " + message.substring(0, Math.min(50, message.length())) + "...");
                        }
                    }
                }
            }
        }
    }
    
    private boolean isPaymentSms(String sender, String message) {
        String senderUpper = sender.toUpperCase();
        String messageUpper = message.toUpperCase();
        
        // Check if sender is from UPI apps
        for (String upiSender : UPI_SENDERS) {
            if (senderUpper.contains(upiSender)) {
                Log.d(TAG, "UPI sender detected: " + sender);
                return true;
            }
        }
        
        // Check if message contains UPI keywords
        for (String keyword : UPI_KEYWORDS) {
            if (messageUpper.contains(keyword.toUpperCase())) {
                Log.d(TAG, "UPI keyword detected: " + keyword);
                return true;
            }
        }
        
        // Check for rupee symbol or amount patterns
        if (messageUpper.contains("₹") || messageUpper.contains("RS.") || messageUpper.contains("INR")) {
            Log.d(TAG, "Currency detected in SMS");
            return true;
        }
        
        return false;
    }
    
    private void sendSmsToFlutter(String sender, String message) {
        Map<String, String> data = new HashMap<>();
        data.put("sender", sender);
        data.put("message", message);
        data.put("text", message);
        data.put("timestamp", String.valueOf(System.currentTimeMillis()));
        
        Log.d(TAG, "Sending SMS to Flutter: " + message);
        
        try {
            MainActivity.sendNotificationEvent(data);
            Log.d(TAG, "SMS sent to Flutter successfully");
        } catch (Exception e) {
            Log.e(TAG, "Error sending SMS to Flutter", e);
        }
    }
    
    private void showNotification(Context context, String title, String message) {
        try {
            // Create intent for notification tap
            Intent intent = new Intent(context, MainActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            
            PendingIntent pendingIntent;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_IMMUTABLE);
            } else {
                pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
            }
            
            // Get default notification sound
            Uri soundUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);
            
            // Build notification with sound
            NotificationCompat.Builder builder = new NotificationCompat.Builder(context, "payment_alerts")
                .setSmallIcon(android.R.drawable.ic_dialog_info)
                .setContentTitle(title)
                .setContentText(message)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setDefaults(NotificationCompat.DEFAULT_ALL)
                .setSound(soundUri)
                .setVibrate(new long[]{0, 500, 1000})
                .setAutoCancel(true)
                .setContentIntent(pendingIntent);
            
            // Show notification
            NotificationManagerCompat notificationManager = NotificationManagerCompat.from(context);
            if (notificationManager.areNotificationsEnabled()) {
                notificationManager.notify((int) System.currentTimeMillis(), builder.build());
                Log.d(TAG, "Notification shown with sound");
            }
        } catch (Exception e) {
            Log.e(TAG, "Error showing notification", e);
        }
    }
}

File: alert_app/android/app/src/main/kotlin/com/example/alert_new/SmsReceiver.kt
package com.example.alert_new

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.telephony.SmsMessage
import android.util.Log

class SmsReceiver : BroadcastReceiver() {
    
    companion object {
        private const val TAG = "SmsReceiver"
        var eventSink: io.flutter.plugin.common.EventChannel.EventSink? = null
    }

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action == "android.provider.Telephony.SMS_RECEIVED") {
            val bundle = intent.extras
            if (bundle != null) {
                val pdus = bundle.get("pdus") as Array<*>?
                if (pdus != null) {
                    for (pdu in pdus) {
                        val sms = SmsMessage.createFromPdu(pdu as ByteArray)
                        val sender = sms.originatingAddress
                        val message = sms.messageBody
                        
                        Log.d(TAG, "SMS from: $sender, Message: $message")
                        
                        // Check if SMS is from payment apps
                        if (isPaymentSms(sender, message)) {
                            val paymentInfo = extractPaymentFromSms(sender, message)
                            if (paymentInfo != null) {
                                eventSink?.success(paymentInfo)
                            }
                        }
                    }
                }
            }
        }
    }
    
    private fun isPaymentSms(sender: String?, message: String): Boolean {
        if (sender == null) return false
        
        val paymentSenders = listOf(
            "GPAY", "GOOGLEPAY", "PHONEPE", "PAYTM", "BHIMUPI", 
            "AMAZONPAY", "MOBIKWIK", "FREECHARGE", "AXISBANK", 
            "SBIUPI", "ICICIBANK", "HDFCBANK"
        )
        
        val upperSender = sender.uppercase()
        val upperMessage = message.uppercase()
        
        return paymentSenders.any { upperSender.contains(it) } &&
               (upperMessage.contains("RECEIVED") || 
                upperMessage.contains("CREDITED") || 
                upperMessage.contains("PAYMENT") ||
                upperMessage.contains("UPI"))
    }
    
    private fun extractPaymentFromSms(sender: String?, message: String): Map<String, Any>? {
        try {
            val upperMessage = message.uppercase()
            
            // Extract amount
            val amountRegex = """RS\.?\s*(\d+(?:,\d+)*(?:\.\d{2})?)""".toRegex()
            val amountMatch = amountRegex.find(upperMessage)
            val amount = amountMatch?.let { "₹${it.groupValues[1]}" } ?: return null
            
            // Extract UPI ID or account info
            val upiRegex = """(\w+@\w+)""".toRegex()
            val upiMatch = upiRegex.find(message)
            val upiId = upiMatch?.value ?: "unknown@upi"
            
            // Determine app name from sender
            val appName = when {
                sender?.contains("GPAY", true) == true -> "Google Pay"
                sender?.contains("PHONEPE", true) == true -> "PhonePe"
                sender?.contains("PAYTM", true) == true -> "Paytm"
                sender?.contains("BHIM", true) == true -> "BHIM UPI"
                sender?.contains("AMAZON", true) == true -> "Amazon Pay"
                else -> "UPI App"
            }
            
            return mapOf(
                "amount" to amount,
                "appName" to appName,
                "upiId" to upiId,
                "text" to message,
                "sender" to (sender ?: ""),
                "timestamp" to System.currentTimeMillis()
            )
        } catch (e: Exception) {
            Log.e(TAG, "Error extracting payment from SMS", e)
            return null
        }
    }
}